### PR TITLE
Redis serialization

### PR DESCRIPTION
**Pull request by ivasio**

### CODE REVIEW

File: aioredisqueue/queue.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 class Queue(object):
 
     def __init__(self, redis, key_prefix='aioredisqueue_', *,
+                 serialization_method=None,
```
Comment (code-of-kpp) [inline aioredisqueue/queue.py]: * let's add ability to use `dynamic` method.

Comment (code-of-kpp) [reply aioredisqueue/queue.py]: let's set default for `key_prefix` to None: and set it dynamically to if this param is None to 'aioredisqueue_' + serialization_method

File: aioredisqueue/task.py [hunk 2]
```diff
@@ -16,8 +16,11 @@ def generate_id(payload=None):
         return str(uuid.uuid4())
 
     @staticmethod
-    def format_payload(payload):
-        return payload
+    def format_payload(payload, serializer='raw'):
+        if serializer == 'raw':
+            return payload
+        else:
+            raise NotImplementedError
```
Comment (code-of-kpp) [inline aioredisqueue/task.py]: Ideally it would be great to have it the way kombu implements it:
https://kombu.readthedocs.io/en/stable/userguide/serialization.html#creating-extensions-using-setuptools-entry-points

but at least we can just create a `serialization` method with `get_serializer()` function which is basically some `dict.__getitem__`

**Review by code-of-kpp — CHANGES_REQUESTED** [review]

File: aioredisqueue/task.py [hunk 3]
```diff
@@ -12,12 +15,32 @@ def generate_ack_info():
...
-        return payload
+    def encode_payload(payload, serialization_method):
+        print(serializers)
+        if serialization_method == 'raw':
+            return bytes(payload.encode(ENCODING))
+        elif serialization_method in serializers.keys():
+            dump = serializers[serialization_method]['dump']
+            return dump(payload)
+        else:
+            raise NotImplementedError
```
Comment (code-of-kpp) [inline aioredisqueue/task.py]: Please run a pylint. You are trying to raise a class instead of an instance of exception.

File: setup.py [hunk 4]
```diff
@@ -11,6 +11,7 @@
     packages=find_packages(),
     install_requires=(
         'aioredis',
+        'pyyaml'
```
Comment (code-of-kpp) [inline setup.py]: it shouldn't be in dependencies

File: test/test_serializers.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+		queue = aioredisqueue.queue.Queue(r, serialization_method='crap')
+		await queue.put(message)
+		task = await queue.get()
+
+
+@pytest.mark.asyncio
+async def test_queue_specified_serializers():
+	r = await aioredis.create_redis(('localhost', 6379), db=0)
+	message = {'a' : 1, '2' : ['b' , [13]]}
+	for method in aioredisqueue.serializers.serializers:
```
Comment (code-of-kpp) [inline test/test_serializers.py]: that should be in pytest.mark.parametrize

File: test/test_task_usage.py [hunk 6]
```diff
@@ -9,7 +9,7 @@
 	r = await aioredis.create_redis(('localhost', 6379), db=0)
 	queue = aioredisqueue.queue.Queue(r)
 	
-	message = b'payload'
+	message = 'payload string 1'
```
Comment (code-of-kpp) [inline test/test_task_usage.py:12]: Try to work with data like this: 
```python
{
    'funny key 🕷': 3,
   b'funny\0\1 key': '🦖'
}
```

File: aioredisqueue/task.py [hunk 7]
```diff
@@ -12,12 +15,32 @@ def generate_ack_info():
         return int(timeit.default_timer() * 1000)
 
     @staticmethod
-    def generate_id(payload=None):
-        return str(uuid.uuid4())
+    def generate_id(serialization_method):
+        return str(uuid.uuid4()) + '_' + serialization_method
```
Comment (code-of-kpp) [inline aioredisqueue/task.py:17]: I'm not sure if we want `serialization_method` in task_id when we have homogeneous queue

File: aioredisqueue/task.py [hunk 8]
```diff
@@ -1,7 +1,10 @@
 import timeit
 import uuid
+from .serializers import serializers
 
 
+ENCODING = 'utf-8'
```
Comment (code-of-kpp) [inline aioredisqueue/task.py]: why is that here?

File: aioredisqueue/serializers.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    'load' : yaml_load,
+    'dump' : yaml_dump
+}})
+
+
+try:
+	import pkg_resources
+
+	for ep in pkg_resources.iter_entry_points(group='serializers'):
+	   serializers.update({ep.name: ep.load()})
```
Comment (code-of-kpp) [inline aioredisqueue/serializers.py]: there isn't much sense in using pkg_resources if resources are not provided in setup.py.

File: aioredisqueue/queue.py [hunk 10]
```diff
@@ -14,6 +15,9 @@ def __init__(self, redis, key_prefix='aioredisqueue_', *,
                  lua_sha=None,
                  loop=None):
 
+        if key_prefix is None:
+            key_prefix = 'aioredisqueue_' + serialization_method
```
Comment (code-of-kpp) [inline aioredisqueue/queue.py:32]: the cases of `serialization_method` being  `None` (for `raw`) or `dynamic`  should be handled and documented. I'm not sure if we need `serialization_method` in queue key if we'll have it in task_ids.

**Review by code-of-kpp — CHANGES_REQUESTED** [review]
CI problems are there because in python3.5 you can't parse json from bytes, you need to decode it first. So for python less than 3.6 you need to do that.
Also please be aware of ujson and other popular alternatives to json module.

File: setup.py (hunk 4 above)
Comment (ivasio) [reply setup.py]: Why? Don't we want to support yaml by default? In such a case pyyaml seems to be right dependency

File: aioredisqueue/task.py (hunk 7 above)
Comment (ivasio) [reply aioredisqueue/task.py:17]: It appears to be a free flexible option and it doesn't make things worse, so why not?

File: setup.py (hunk 4 above)
Comment (code-of-kpp) [reply setup.py]: Because we don't want people to force installing packages for every serialization method we support.

File: aioredisqueue/task.py (hunk 7 above)
Comment (code-of-kpp) [reply aioredisqueue/task.py:17]: I think it is just a matter of taste. Avoiding data duplication seems to be the right thing to do

File: setup.py (hunk 4 above)
Comment (ivasio) [reply setup.py]: So should we leave it up to users to specify this serializer via setuptools in this case? As far as I understand, support means guarantee that it works ok and install required external packages if situation needs it

File: test/test_task_usage.py (hunk 6 above)
Comment (ivasio) [reply test/test_task_usage.py:12]: Default json module doesn't support binary keys in dicts to serialize. Seems like I should I use an alternative to default json module, should I specify it in dependencies?

File: aioredisqueue/queue.py (hunk 10 above)
Comment (ivasio) [reply aioredisqueue/queue.py:32]: The reason for having it as `queue` constructor parameter is it just provides a way to specify the default encoder different from `json` once instead of every time when putting a new task. Yes, having it at the same time as Redis key is actually data duplication, but it provides the easiest usage for the very popular cases of using json or custom loader just by specifying it once in each producer

File: aioredisqueue/serializers.py (hunk 9 above)
Comment (ivasio) [reply aioredisqueue/serializers.py]: I haven't found a good way to parse `setup()` parameters without using `pkg_resources`. Could you give an advice of any?

File: setup.py (hunk 4 above)
Comment (code-of-kpp) [reply setup.py]: Usually it goes like this: if user specifies serialization method known to the library but corresponding library (like `yaml`) is not installed they get `ImportError`. When they try to use serialization method not known to the library (i.e. no corresponding `entry_point` exists) they got our `UnknownSerializationMethod` error.
Additional dependencies for known serialization methods (like yaml or MessagePack) can be provided as `extra_requires` in `setup.py`.

File: test/test_task_usage.py (hunk 6 above)
Comment (code-of-kpp) [reply test/test_task_usage.py:12]: I'm actually making a point here against JSON as a default serialization method.
try this:
```JSON
{
     1:  "one",
     2: "two",
     "1": "also one"
}
```

or this:

```python
{
    ('tuple', 'key'): ['list', 'values'],
    frozenset('set'): ('tuple', 'values'),
}
```

File: aioredisqueue/serializers.py (hunk 9 above)
Comment (code-of-kpp) [reply aioredisqueue/serializers.py]: I'm just saying that `entry_points` were not specified in `setup.py`

File: aioredisqueue/queue.py (hunk 10 above)
Comment (code-of-kpp) [reply aioredisqueue/queue.py:32]: Yeah, maybe, but if I specify `serialization_method='raw'` or (which is meant to be the same) `serialization_method=None`,  `key_prefix` shouldn't end with `_raw` and with `None` you'll have an error here.

File: aioredisqueue/serializers.py [hunk 11]
```diff
@@ -1,35 +1,42 @@
...
+        'load': lambda payload: payload.decode('utf-8'),
+        'dump': lambda message: bytes(message.encode('utf-8'))
+    }
 }
 
 
-def yaml_load(arg):
-    return yaml.load(arg, Loader=yaml.Loader)
+try:
+    import yaml
```
Comment (code-of-kpp) [inline aioredisqueue/serializers.py]: That should be lazy.

File: test/test_task_usage.py (hunk 6 above)
Comment (ivasio) [reply test/test_task_usage.py:12]: Set `pickle` to be default

File: aioredisqueue/task.py (hunk 7 above)
Comment (ivasio) [reply aioredisqueue/task.py:17]: Yes, there is some data duplication, but I decided to leave it as tradeoff to maximal flexibility and interoperability of homogenous and heterogenous queues

File: aioredisqueue/serializers.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        'load': lambda payload: json.loads(payload.decode('utf-8')),
+        'dump': json.dumps
+    },
+    'pickle': {
+        'load': pickle.loads,
+        'dump': pickle.dumps
+    },
+    'raw': {
+        'load': lambda payload: payload.decode('utf-8'),
+        'dump': lambda message: bytes(message.encode('utf-8'))
```
Comment (code-of-kpp) [inline aioredisqueue/serializers.py]: `Raw` should work with bytes directly - avoiding encode/decode

File: aioredisqueue/serializers.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
+import json
+import pickle
+
+_serializers = {
+    'json': {
+        'load': lambda payload: json.loads(payload.decode('utf-8')),
```
Comment (code-of-kpp) [inline aioredisqueue/serializers.py:6]: Decode part here should be a developer (user) responsibility

**Comment by code-of-kpp** [comment]
#6

File: aioredisqueue/task.py (hunk 7 above)
Comment (code-of-kpp) [reply aioredisqueue/task.py:17]: So do we have solution for that? If it is decided to postpone it there should be a link to corresponding issue posted in this thread.

File: aioredisqueue/queue.py [hunk 14]
```diff
@@ -4,15 +4,36 @@
 
 
 class Queue(object):
-
-    def __init__(self, redis, key_prefix='aioredisqueue_', *,
```
Comment (code-of-kpp) [inline aioredisqueue/queue.py:8]: Looks like you made keyword-only arguments position dependent.

File: aioredisqueue/task.py [hunk 15]
```diff
@@ -12,13 +13,21 @@ def generate_ack_info():
...
-        return payload
-
+    def encode_payload(payload, serialization_method):
+        dump = get_serializer(serialization_method)['dump']
+        return dump(payload)
+    
+    @staticmethod
+    def decode_payload(task_id, payload):
+        task_id = task_id.decode('utf-8')
+        serialization_method = task_id[task_id.find('_') + 1:]
```
Comment (code-of-kpp) [inline aioredisqueue/task.py]: what if task_id doesn't contain `_` and `serialization method`? (current behavior)

**Review by code-of-kpp — CHANGES_REQUESTED** [review]

File: aioredisqueue/task.py (hunk 7 above)
Comment (ivasio) [reply aioredisqueue/task.py:17]: I don't understand why did "the matter of taste" become something that must be fixed anyway. I explained why did I decide to leave this duplication, what's the problem?

File: aioredisqueue/task.py (hunk 15 above)
Comment (ivasio) [reply aioredisqueue/task.py]: Current behavior is opposite, `task_id` does contian `_` and `serialization method`. It is guaranteed by `Task.generate_id()`

File: aioredisqueue/queue.py (hunk 14 above)
Comment (ivasio) [reply aioredisqueue/queue.py:8]: It did not conform to my `autopep`, it didn't contain this rule. Will fix it

File: aioredisqueue/task.py (hunk 7 above)
Comment (code-of-kpp) [reply aioredisqueue/task.py:17]: Well, we just don't have a consensus here. It doesn't mean that PR wouldn't be merged but if someone comes across the same ideas they wouldn'tfind this discussion.So if you are not going to change this, we need the discussion to continue in separate issue (which doesn't mean that it "must be fixed anyway".)

File: aioredisqueue/queue.py (hunk 14 above)
Comment (code-of-kpp) [reply aioredisqueue/queue.py:8]: This is out of scope of autopep (or even and better - `black`). But if you have a lot of arguments and some of them may change in the future, allowing passing arguments as positional could be harmful.

File: aioredisqueue/task.py (hunk 15 above)
Comment (code-of-kpp) [reply aioredisqueue/task.py]: Currently in `master` we have:

```python
    @staticmethod
    def generate_id(payload=None):
        return str(uuid.uuid4())

Comment (ivasio) [reply aioredisqueue/task.py]: This function is intended to be merged into `master` with following version of `generate_id` :
```python
    def generate_id(serialization_method):
        return str(uuid.uuid4()) + '_' + serialization_method
```

File: aioredisqueue/task.py (hunk 7 above)
Comment (ivasio) [reply aioredisqueue/task.py:17]: Moved this discussion to #13

File: aioredisqueue/task.py (hunk 15 above)
Comment (code-of-kpp) [reply aioredisqueue/task.py]: So to update queue to new version, the tasks must be drained? That should be fixed or at least documented as backward-incompatible change.
